# Typing Speeds

How can you improve your typing speed?

The file `typing-speeds.csv` contains typing speed data from >168,000 people typing 15 sentences each. The data was collected via an online typing test published at a free typing speed assessment webpage.

In [1]:
# FOR GOOGLE COLAB ONLY.
# Uncomment and run the code below. A dialog will appear to upload files.
# Upload 'typing-speeds.csv'.

# from google.colab import files
# uploaded = files.upload()

In [2]:
import pandas as pd

df = pd.read_csv('typing-speeds.csv')
df

,PARTICIPANT_ID,AGE,HAS_TAKEN_TYPING_COURSE,COUNTRY,LAYOUT,NATIVE_LANGUAGE,FINGERS,KEYBOARD_TYPE,ERROR_RATE,AVG_WPM_15,ROR
0,3,30,0,US,qwerty,en,1-2,full,0.511945,61.9483,0.2288
1,5,27,0,MY,qwerty,en,7-8,laptop,0.871080,72.8871,0.3675
2,7,13,0,AU,qwerty,en,7-8,laptop,6.685633,24.1809,0.0667
3,23,21,0,IN,qwerty,en,3-4,full,2.130493,24.7112,0.0413
4,24,21,0,PH,qwerty,tl,7-8,laptop,1.893287,45.3364,0.2678
...,...,...,...,...,...,...,...,...,...,...,...
168589,517932,20,0,US,qwerty,en,9-10,laptop,8.731466,24.9125,0.1842
168590,517936,25,0,PL,qwerty,pl,9-10,laptop,0.000000,66.2946,0.0639
168591,517943,38,1,US,qwerty,en,9-10,laptop,0.147929,75.6713,0.2021
168592,517944,28,0,GB,qwerty,en,9-10,laptop,0.278552,91.7083,0.5133




| **Variable**             | **Description**                                                                 |
|--------------------------|---------------------------------------------------------------------------------|
| `PARTICIPANT_ID`         | Unique ID of the participant                                                   |
| `AGE`                    | Age of the participant                                                         |
| `HAS_TAKEN_TYPING_COURSE`| Whether the participant has taken a typing course (1 = Yes, 0 = No)            |
| `COUNTRY`                | Country of the participant                                                     |
| `LAYOUT`      		   | Keyboard layout used (QWERTY, AZERTY, or QWERTZ)                               |
| `NATIVE_LANGUAGE`        | Native language of the participant                                             |
| `FINGERS`                | Number of fingers used for typing (options: 1-2, 3-4, 5-6, 7-8, 9-10)          |
| `KEYBOARD_TYPE`          | Type of keyboard used (Full/desktop, laptop, small physical, or touch)         |
| `ERROR_RATE(%)`          | Uncorrected error rate (as a percentage)                                       |
| `AVG_WPM_15`             | Words per minute averaged over 15 typed sentences                              |
| `ROR`                    | Rollover ratio                                                                 |


### Project Ideas
- Remove unnecessary columns, such as PARTICIPANT_ID, to streamline the dataset.

- Rename columns (e.g `AVG_WPM_15` to `wpm`, `ROR` to `ror`, `HAS_TAKEN_TYPING_COURSE` to `course`) for brevity and clarity during analysis.

Finger Count Analysis
- Compare typing speeds across groups using different numbers of fingers, excluding the "10+" category for simplicity.

- Control for consistency by first filtering to similar `AGE`, `KEYBOARD_LAYOUT`, `NATIVE_LANGUAGE`, `KEYBOARD_TYPE`, and `HAS_TAKEN_TYPING_COURSE` values.

- Exclude participants with high error rates (ERROR_RATE > 3%) to focus on reliable data.

- Drop columns after filtering if they now only have a single value.

Rollover Ratio Analysis
- The Rollover Ratio (`ROR`) represents the proportion of keypresses where a new key is pressed before releasing the previous one.

- Compare typing speeds between participants with `ROR` ≤ 20% and those with `ROR` > 80%, keeping `AGE`, `KEYBOARD_TYPE`, `FINGERS`, and other variables constant.

Influence of Typing Course
- Compare typing speeds between participants with a typing course (`HAS_TAKEN_TYPING_COURSE` = 1) and without (`HAS_TAKEN_TYPING_COURSE` = 0), holding other variables such as `KEYBOARD_TYPE`, `AGE` range, and `FINGER_COUNT` constant.


# PROJECT IDEAS

**1. Remove unnecessary columns, such as PARTICIPANT_ID, to streamline the dataset.**

In [3]:
#creating a copy of the original df (typing)
typing = df.copy()

#removing PARTICIPANT_ID:
typing = typing.drop('PARTICIPANT_ID', axis=1) #axis=1 means 'column', 0 is for rows

**2. Rename columns (e.g `AVG_WPM_15` to `wpm`, `ROR` to `ror`, `HAS_TAKEN_TYPING_COURSE` to `course`) for brevity and clarity during analysis.**

In [4]:
name_map = {
    'AGE' : 'age',
    'HAS_TAKEN_TYPING_COURSE' : 'course',
    'COUNTRY' : 'country',
    'LAYOUT' : 'layout',
    'NATIVE_LANGUAGE' : 'language',
    'FINGERS' : 'fingers',
    'KEYBOARD_TYPE' : 'keyboard',
    'ERROR_RATE' : 'error_rate',
    'AVG_WPM_15' : 'wpm',
    'ROR' : 'ror'
}

typing = typing.rename(columns = name_map)
typing


,age,course,country,layout,language,fingers,keyboard,error_rate,wpm,ror
0,30,0,US,qwerty,en,1-2,full,0.511945,61.9483,0.2288
1,27,0,MY,qwerty,en,7-8,laptop,0.871080,72.8871,0.3675
2,13,0,AU,qwerty,en,7-8,laptop,6.685633,24.1809,0.0667
3,21,0,IN,qwerty,en,3-4,full,2.130493,24.7112,0.0413
4,21,0,PH,qwerty,tl,7-8,laptop,1.893287,45.3364,0.2678
...,...,...,...,...,...,...,...,...,...,...
168589,20,0,US,qwerty,en,9-10,laptop,8.731466,24.9125,0.1842
168590,25,0,PL,qwerty,pl,9-10,laptop,0.000000,66.2946,0.0639
168591,38,1,US,qwerty,en,9-10,laptop,0.147929,75.6713,0.2021
168592,28,0,GB,qwerty,en,9-10,laptop,0.278552,91.7083,0.5133


# FINGER COUNT ANALYSIS

**Finger Count Analysis 1: Compare typing speeds across groups using different numbers of fingers, excluding the "10+" category for simplicity.**

In [5]:
#dropping rows with 10+ fingers
no_11 = typing.query('fingers != "10+"')

#comparing the average typing speeds across groups using different numbers of fingers.
no_11.groupby('fingers')['wpm'].mean()

fingers
1-2     40.280812
3-4     41.004952
5-6     45.731789
7-8     50.057909
9-10    57.379572
Name: wpm, dtype: float64

**Inference:** The result implies that people who use 9-10 fingers types the fastest in general. **But**, this method only groups the participants by number of fingers they use, disregarding the other factors. Meaning, people in a specific group may have different age, keyboard layouts, keyboard types, language, and may or may not have taken typing course. Let's address this... 

**Finger Count Analysis 2: Control for consistency by first filtering to similar `AGE`, `KEYBOARD_LAYOUT`, `NATIVE_LANGUAGE`, `KEYBOARD_TYPE`, and `HAS_TAKEN_TYPING_COURSE` values.**

**For this analysis, let's use the group with the ff criteria:**
- age: 25-59 (adults)
- keyboard_layout: qwerty
- native_language: en
- keyboard_type: laptop
- course: 0 (no)

In [6]:
typing_grp = no_11.query(
    'age >=25 and age <= 59 and '
    'layout == "qwerty" and '
    'language == "en" and '
    'keyboard == "laptop" and '
    'course == 0'
)

typing_grp

,age,course,country,layout,language,fingers,keyboard,error_rate,wpm,ror
1,27,0,MY,qwerty,en,7-8,laptop,0.871080,72.8871,0.3675
10,34,0,PH,qwerty,en,5-6,laptop,1.667930,39.2606,0.0867
13,25,0,US,qwerty,en,1-2,laptop,3.183792,28.1308,0.1019
56,25,0,SG,qwerty,en,5-6,laptop,0.369004,90.8364,0.6011
70,30,0,IN,qwerty,en,7-8,laptop,0.793651,31.5854,0.0482
...,...,...,...,...,...,...,...,...,...,...
168570,32,0,CA,qwerty,en,3-4,laptop,0.357143,23.6640,0.0475
168571,31,0,IN,qwerty,en,3-4,laptop,0.941915,45.3349,0.0639
168579,38,0,GB,qwerty,en,5-6,laptop,1.551724,41.8037,0.0818
168592,28,0,GB,qwerty,en,9-10,laptop,0.278552,91.7083,0.5133


The original data with 168,594 total number of rows is now filtered into 17,273 rows, representing a specific group of participants: **Adult english speaking participants that use laptop with qwerty keyboard and has not taken any typing course.**

**Finger Count Analysis 3: Exclude participants with high error rates (ERROR_RATE > 3%) to focus on reliable data.**

In [7]:
typing_grp = typing_grp.query('error_rate < 3')
typing_grp

,age,course,country,layout,language,fingers,keyboard,error_rate,wpm,ror
1,27,0,MY,qwerty,en,7-8,laptop,0.871080,72.8871,0.3675
10,34,0,PH,qwerty,en,5-6,laptop,1.667930,39.2606,0.0867
56,25,0,SG,qwerty,en,5-6,laptop,0.369004,90.8364,0.6011
70,30,0,IN,qwerty,en,7-8,laptop,0.793651,31.5854,0.0482
73,31,0,US,qwerty,en,1-2,laptop,0.309119,56.6029,0.3567
...,...,...,...,...,...,...,...,...,...,...
168570,32,0,CA,qwerty,en,3-4,laptop,0.357143,23.6640,0.0475
168571,31,0,IN,qwerty,en,3-4,laptop,0.941915,45.3349,0.0639
168579,38,0,GB,qwerty,en,5-6,laptop,1.551724,41.8037,0.0818
168592,28,0,GB,qwerty,en,9-10,laptop,0.278552,91.7083,0.5133


**Finger Count Analysis 4: Drop columns after filtering if they now only have a single value.**

Since the data now shows single value for course, layout, language, ad keyboard, we can just remove these columns to simplify the table and  focus more on the analysis.

In [8]:
typing_grp = typing_grp.drop(columns = ['course', 'layout', 'language', 'keyboard'])
typing_grp

,age,country,fingers,error_rate,wpm,ror
1,27,MY,7-8,0.871080,72.8871,0.3675
10,34,PH,5-6,1.667930,39.2606,0.0867
56,25,SG,5-6,0.369004,90.8364,0.6011
70,30,IN,7-8,0.793651,31.5854,0.0482
73,31,US,1-2,0.309119,56.6029,0.3567
...,...,...,...,...,...,...
168570,32,CA,3-4,0.357143,23.6640,0.0475
168571,31,IN,3-4,0.941915,45.3349,0.0639
168579,38,GB,5-6,1.551724,41.8037,0.0818
168592,28,GB,9-10,0.278552,91.7083,0.5133


Running another average comparison across groups of participant that use different number of fingers in typing.

In [9]:
typing_grp.groupby('fingers')['wpm'].mean()

fingers
1-2     42.174141
3-4     41.541451
5-6     46.706966
7-8     51.520652
9-10    58.949967
Name: wpm, dtype: float64

**ANSWER:** The results show a similar comparison with the unfiltered data -- People that use 9-10 fingers when typing tend to be faster than those who use fewer fingers. (This takes the other factors into consideration, making the inference and interpretation more specific.)

# ROLLOVER RATIO ANALYSIS

**The Rollover Ratio (`ROR`) represents the proportion of keypresses where a new key is pressed before releasing the previous one.**

**Rollover Ratio Analysis: Compare typing speeds between participants with `ROR` ≤ 20% and those with `ROR` > 80%, keeping `AGE`, `KEYBOARD_TYPE`, `FINGERS`, and other variables constant.**

**For this analysis, let's use the group with the ff criteria:**
- age: 25-59 (adults)
- keyboard: laptop
- fingers: 9-10
- course: 0 (no)
- layout: qwerty
- language: en
- error_rate: < 3.00


In [21]:
# filtering the dataset:
rra = typing.query(
    'age >=25 and age <= 59 and '
    'keyboard == "laptop" and '
    'fingers == "9-10" and ' 
    'course == 0 and '
    'layout == "qwerty" and '
    'language == "en" and '
    'error_rate < 3'
)
rra

,age,course,country,layout,language,fingers,keyboard,error_rate,wpm,ror
74,27,0,GB,qwerty,en,9-10,laptop,1.791531,97.7603,0.4686
94,26,0,PH,qwerty,en,9-10,laptop,0.523560,83.9777,0.6709
95,35,0,IN,qwerty,en,9-10,laptop,1.018330,67.3202,0.4596
120,32,0,GB,qwerty,en,9-10,laptop,0.310559,71.1093,0.5973
159,30,0,IN,qwerty,en,9-10,laptop,0.619195,19.3110,0.0897
...,...,...,...,...,...,...,...,...,...,...
168475,35,0,SG,qwerty,en,9-10,laptop,1.257862,70.9405,0.2616
168489,31,0,US,qwerty,en,9-10,laptop,0.561010,40.4361,0.3125
168505,38,0,PK,qwerty,en,9-10,laptop,0.510204,79.8501,0.5374
168517,25,0,US,qwerty,en,9-10,laptop,0.754148,52.5135,0.2892


In [36]:
#grouping the filtered dataset and performing rra
#we can't use groupby, so let's use query
uppergrp_mean = rra.query('ror > 0.8')['wpm'].mean()
uppergrp_mean

np.float64(104.76685714285715)

In [27]:
lowergrp_mean = rra.query('ror <= 0.2')['wpm'].mean()
lowergrp_mean

np.float64(41.68487653319818)

**ANSWER:** The results indicate that adult participants in a specific group (those who are english native speakers that use a laptop with qwerty keyboard, use 9-10 fingers in typing, with less than 3% error rate, and has not taken any typing course) differs in typing speeds when grouped according to ROR (Rollover Ratio). People who has a ROR value of greater than  80% has a group wpm mean of 104.76. Whereas, people with that of less than or equal to 20% has a wpm average of 41.68. **This means that people with high Rollover Ratio (ROR), or high proportion of new keypress before releasing the previous one, has a higher word per minute (WPM) in average compared to those with low ROR.**

# INFLUENCE OF TYPING COURSE

**Influence of Typing Course: Compare typing speeds between participants with a typing course (`HAS_TAKEN_TYPING_COURSE` = 1) and without (`HAS_TAKEN_TYPING_COURSE` = 0), holding other variables such as `KEYBOARD_TYPE`, `AGE` range, and `FINGER_COUNT` constant.**

**For this analysis, let's use the group with the ff criteria:**
- age: 25-59 (adults)
- keyboard: laptop
- fingers: 9-10
- layout: qwerty
- language: en
- error_rate: < 3.00

In [31]:
itc = typing.query(
    'age >=25 and age <= 59 and '
    'keyboard == "laptop" and '
    'fingers == "9-10" and '
    'layout == "qwerty" and '
    'language == "en" and '
    'error_rate < 3'
)
itc

,age,course,country,layout,language,fingers,keyboard,error_rate,wpm,ror
60,50,1,US,qwerty,en,9-10,laptop,1.136364,53.6687,0.1722
74,27,0,GB,qwerty,en,9-10,laptop,1.791531,97.7603,0.4686
94,26,0,PH,qwerty,en,9-10,laptop,0.523560,83.9777,0.6709
95,35,0,IN,qwerty,en,9-10,laptop,1.018330,67.3202,0.4596
120,32,0,GB,qwerty,en,9-10,laptop,0.310559,71.1093,0.5973
...,...,...,...,...,...,...,...,...,...,...
168540,48,1,AU,qwerty,en,9-10,laptop,1.733809,69.1523,0.1625
168557,55,1,US,qwerty,en,9-10,laptop,2.427184,44.0643,0.2255
168569,32,1,PH,qwerty,en,9-10,laptop,0.172117,36.0377,0.0608
168591,38,1,US,qwerty,en,9-10,laptop,0.147929,75.6713,0.2021


In [34]:
#grouping the filtered dataset and comparing the wpm
itc_mean = itc.groupby('course')['wpm'].mean()
itc_mean

course
0    58.949967
1    62.908945
Name: wpm, dtype: float64

**ANSWER:** The average word per minute (WPM) of the groups of people who has taken a typing course and those who have not states that people with prior typing training types faster, with an average of 62.91 wpm, than people without, which has an wpm average of 58.95. 